CHANGELOG.md

```diff
@@ -1,54 +1,126 @@
-  * `st/decode` to transform and validate value from external format into valid value defined by the spec
-  * `st/encode` to transform (without validation) a value into external format
-  * new `encode` and `decode` namespaces to define transformations for `Spec`s. `json` and `string` are supported by default, but free to extend whatever (`xml`, `avro` etc.)
-  * type-based encoders can be defined via `st/type-conforming` (defaults no none)
+* new `st/IntoSpec` protocol to convert non-recursively vanilla `clojure.spec` Specs into `st/Spec`:s. Used in `st/encode`, `st/decode`, `st/explain`, `st/explain-data`, `st/conform` and `st/conform!`.
+* **BREAKING**: Bye bye conforming, welcome transformers!
+  * Guide: https://github.com/metosin/spec-tools#spec-driven-transformations
+  * removed: `st/type-conforming`, `st/json-conforming`, `st/string-conforming`
+  * new `st/Transformer` protocol to drive spec-driven value transformations
+  * spec values can be both encoded (`st/encode`) & decoded (`st/decode`) using a transformer, fixes [#96](https://github.com/metosin/spec-tools/issues/96).
+  * renamed ns `spec-tools.conform` into `spec-tools.transform`, covering both encoding & decoding of values
+  * `st/type-transformer`, supporting both `:type` and `Spec` level transformations
+  * Spec-driven transformations via keys in `encode` and `decode` namespaces.
+  * `st/encode`, `st/decode`, `st/explain`, `st/explain-data`, `st/conform` and `st/conform!` take the transformer instance an optional third argument
+     * `st/json-transformer`, `st/string-transformer`, `strip-extra-keys-transformer` and `fail-on-extra-keys-transformer` are shipped out-of-the-box.
+
+### Transformer
+
+```clj
+(defprotocol Transformer
+  (-name [this])
+  (-encoder [this spec value])
+  (-decoder [this spec value]))
+```
+
+
+### Spec-driven transformations
+
+* use `:encode/*` and `:decode/*` keys from Spec instances to declare how the values should be transformed
 
 ```clj
 (require '[clojure.spec.alpha :as s])
 (require '[spec-tools.core :as st])
+(require '[clojure.string :as str])
 
 (s/def ::spec
   (st/spec
     {:spec #(and (simple-keyword? %) (-> % name str/lower-case keyword (= %)))
-     :description "a lowercase simple keyword, encoded in uppercase in string-mode"
+     :description "a lowercase keyword, encoded in uppercase in string-mode"
      :decode/string #(-> %2 name str/lower-case keyword)
      :encode/string #(-> %2 name str/upper-case)}))
 
-; decode also validates
-(st/decode ::spec "kikka")
-; => :clojure.spec.alpha/invalid
+(st/decode ::spec :kikka)
+; :kikka
 
-(st/decode ::spec "kikka" st/string-conforming)
-; => :kikka
+(as-> "KiKka" $
+      (st/decode ::spec $))
+; :clojure.spec.alpha/invalid
+
+(as-> "KiKka" $
+      (st/decode ::spec $ st/string-transformer))
+; :kikka
+
+(as-> "KiKka" $
+      (st/decode ::spec $ st/string-transformer)
+      (st/encode ::spec $))
```

Comment:
I'm not sure what is supposed to happen here. If `(st/encode ::spec $)` won't ever do anything useful, should it have a 3-ary version at all?

Reply:
Right, I meant 2, not 3